aten/src/ATen/mps/MPSAllocator.h

```diff
@@ -0,0 +1,242 @@
+      return [heap maxAvailableSizeWithAlignment:Alignment];
+  }
+  id<MTLBuffer> newMTLBuffer(size_t length, bool is_shared) {
+    id<MTLBuffer> buf = [heap newBufferWithLength:length options:getOptions(is_shared)];
+    if (buf) {
+      size.available = heapAvailableSize(heap);
+      n_buffers++;
+    }
+    return buf;
+  }
+  void releaseMTLBuffer(id<MTLBuffer> buffer) {
+    [buffer release];
+    size.available = heapAvailableSize(heap);
+    n_buffers--;
+  }
+  void releaseMTLHeap() {
+    TORCH_INTERNAL_ASSERT(!n_buffers); // assert if heap isn't empty
+    [heap release];
+    size.available = 0;
+  }
+};
+typedef bool (*HeapComparison)(const HeapBlock*, const HeapBlock*);
+
+struct BufferPool
+{
+  BufferPool(const id<MTLDevice> Device, bool Small, bool Shared) :
+           device(Device), is_small(Small), is_shared(Shared),
+           heaps(HeapBlock::Comparator), buffers(BufferBlock::Comparator) { }
+
+  const id<MTLDevice> device;
+  // small heaps have sizes of kSmallHeap, and large ones kLargeHeap
+  const bool is_small;
+  // private pools allocated on device memory; otherwise, shared between host/device
+  const bool is_shared;
+  // list of heaps ordered by their "available" (not total) memory size
+  std::set<HeapBlock*, HeapComparison> heaps;
+  // list of only "available" buffers in the pool (i.e., buffers not in-use)
+  std::set<BufferBlock*, BufferComparison> buffers;
+};
+
+struct AllocParams
+{
+  AllocParams(size_t Alloc_Size, size_t Requested_Size, BufferPool* Pool) :
+            search_key(Alloc_Size), pool(Pool),
+            buffer_block(nullptr), requested_size(Requested_Size) {}
+  size_t size() const { return search_key.size; }
+
+  BufferBlock search_key;
+  BufferPool* pool;
+  BufferBlock* buffer_block;
+  size_t requested_size;
+};
+
+class MPSHeapAllocatorImpl
+{
+public:
+  explicit MPSHeapAllocatorImpl() :
+                      m_device(at::mps::MPSDevice::getInstance()->device()),
+                      m_large_pool_shared(m_device, false, true), m_large_pool_private(m_device, false, false),
+                      m_small_pool_shared(m_device, true , true), m_small_pool_private(m_device, true , false),
+                      m_total_allocated_memory(0), m_max_buffer_size([m_device maxBufferLength]),
+                      m_set_fraction(false), m_enable_debug_info(false) { }
+
+  // interface exposed to at::Allocator
+  id<MTLBuffer> Malloc(size_t size, bool sharedStorage);
+  void Free(void* ptr);
+  void EmptyCache();
+  bool isSharedBuffer(void* ptr);
+
+  inline id<MTLDevice> Device() const { return m_device; }
+  void enable_debug_info() { m_enable_debug_info = true; }
+  bool debug_info_enabled() const { return m_enable_debug_info; }
+  void set_shared_storage_mode(bool useSharedStorage);
+
+private:
+  const id<MTLDevice> m_device;
+  std::mutex m_mutex;
+  // allocated buffers by device pointer
+  std::unordered_map<void*, BufferBlock*> m_allocated_buffers;
```

Comment:
consider using a flat hash map instead

Reply:
What's the motivation to use flat hash map here instead of unordered_map (which also is a STL C++ implementation for hash map).

Reply:
Added flat hash map.